comet-java-client/src/main/java/ml/comet/experiment/impl/BaseExperiment.java

```diff
@@ -0,0 +1,899 @@
+     * @param <T>      the data type to be returned.
+     * @return the loaded data.
+     */
+    private <T> T loadRemote(final Function<String, Single<T>> loadFunc, String alias) {
+        return validateAndGetExperimentKey()
+                .concatMap(loadFunc)
+                .doOnError(ex -> getLogger().error("Failed to read {} for the experiment, experiment key: {}",
+                        alias, this.experimentKey, ex))
+                .blockingGet();
+    }
+
+    /**
+     * Uses provided function to send request data asynchronously and log received output. Optionally, can use
+     * provided {@link Action} handler to notify about completion of the operation.
+     *
+     * @param func       the function to be invoked to send request data.
+     * @param request    the request data object.
+     * @param onComplete the optional {@link Action} to be notified the operation completes either
+     *                   successfully or erroneously.
+     * @param <T>        the type of the request data object.
+     */
+    private <T> void sendAsynchronously(final BiFunction<T, String, Single<LogDataResponse>> func,
+                                        final T request, final Action onComplete) {
+        Single<LogDataResponse> single = validateAndGetExperimentKey()
+                .subscribeOn(Schedulers.io())
+                .concatMap(experimentKey -> func.apply(request, experimentKey));
+
+        // register notification action if provided
+        if (onComplete != null) {
+            single = single.doFinally(onComplete);
+        }
+
+        // subscribe to receive operation results
+        single
+                .observeOn(Schedulers.single())
+                .subscribe(
+                        (logDataResponse) -> DataResponseLogger.checkAndLog(logDataResponse, getLogger(), request),
+                        (throwable) -> getLogger().error("failed to log {}", request, throwable),
+                        disposables);
+    }
+
+    /**
+     * Uses provided function to send request data synchronously. If response indicating the remote error
+     * received the {@link CometApiException} will be thrown.
+     *
+     * @param func    the function to be invoked to send request data.
+     * @param request the request data object.
+     * @param <T>     the type of the request data object.
+     * @throws CometApiException if received response with error indicating that data was not saved.
+     */
+    private <T> void sendSynchronously(final BiFunction<T, String, Single<LogDataResponse>> func,
+                                       final T request) throws CometApiException {
+        LogDataResponse response = validateAndGetExperimentKey()
+                .concatMap(experimentKey -> func.apply(request, experimentKey))
+                .blockingGet();
+
+        if (response.hasFailed()) {
+            throw new CometApiException("Failed to log {}, reason: %s", request, response.getMsg());
```

Comment:
@yaricom - if we log metric (sync) and it will get a non `2xx` response will it throw an exception? we should not fail the run if we fail to log something

Reply:
The OnlineExperiment is using asynchronous version of data logging, which doesn't re-throw exceptions - only log them.

The synchronous versions is intended to be used by other types of experiments, like ApiExperiment for example.